### DISCUSSION TITLE
A new variable for current file's content

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#57] [Ideas - done & released]
@FelipeRearden mentioned an idea about a new variable named `{{NoteContent}}` in [a discussion about making `{{selection}}` variable to select text automatically if nothing is selected](https://github.com/Taitava/obsidian-shellcommands/discussions/54#discussioncomment-1404928).

> This variable was in my mind for some time.
> ![kuva](https://user-images.githubusercontent.com/13002142/135376433-9597094c-fd47-4955-ba61-20d561cca997.png)

I have been thinking about this, too, but haven't so far made a decision about it. One reason why I did not yet see this so important, was the ability to use something like `cat {{file_path:absolute}}` (on Linux) to get access to the current file's content. But this approach has the YAML problem that you Felipe mentioned. So a YAMLless option is needed.

Naming is also one big questionmark to me still. I'm considering the following options, but these are not any kind of decisions yet:
- `{{file_content}}`: Give the whole content, including YAML. Belongs to the same family with `{{file_name}}` and `{{file_path}}`.
- `{{content}}`: Give content without YAML. Belongs to the same family with `{{title}}`.
- `{{yaml_content}}`: Give only YAML. Alternative ideas: `{{frontmatter_content}}`, `{{headers}}`. These would be more agnostic regarding the markup language as they do not mention YAML.
- `{{yaml:variablename}}`: More on this idea in it's own discussion here: https://github.com/Taitava/obsidian-shellcommands/discussions/56 . I just mentioned it here because name collision should be avoided.

**Comment by FelipeRearden** [comment]
My opinion is that: 
- `file` when the variable is related to something outside Obsidian `file_path`
- `note` when the variable is related to something inside Obsidian
- `content` when is something inside the note `yaml_content` and `content` itself

**Reply by Taitava** [reply]
Sorry, I didn't quite understand these. Can you give some examples, please? 🙂

**Reply by FelipeRearden** [reply]
Sorry @Taitava !

Sure, let me rephrase:

> Naming is also one big questionmark to me still. I'm considering the following options, but these are not any kind of decisions yet:

My idea is to use the therms file, content and note based on this criteria:

- The word `file` when the variable is related to something outside Obsidian: example -> `file_path` Instead of `note_path`
- The word `note` when the variable is related to something inside Obsidian: example -> `note_title` Instead of `file_title`
- The `content` when is related with something **inside** the note: example -> `yaml_content` and `note_content` 

:)

**Reply by Taitava** [reply]
> The word `file` when the variable is related to something outside Obsidian: example -> `file_path` Instead of note_path

I do agree with this one. Properties that relate tightly to file*system* deserve the `file_` prefix.

> The word `note` when the variable is related to something inside Obsidian: example -> `note_title` Instead of file_title

This could be logical, but unfortunately it would break the convention that `{{title}}` uses: it doesn't have a `note_` prefix, although it's related to a note. And `{{title}}` name comes from Obsidian, so I would have a steep threshold to change it. Generally, I am really open for discussion about renaming current variables in this stage where the plugin has not yet reached a stable `1.0.0` state, but changing `{{title}}` would require quite a lot of reasoning. 🙂 

And, if we cannot change `{{title}}` to have a `note_` prefix, I'm thinking that maybe we should not have a `note_` prefix in any variable?

>     * The `content` when is related with something **inside** the note: example -> `yaml_content` and `note_content`

Would `note_content` then be explicitly content without YAML? What would be a name for a third variable that would give all content, including YAML?

**Reply by FelipeRearden** [reply]
> Generally, I am really open for discussion about renaming current variables in this stage where the plugin has not yet reached a stable 1.0.0 state, but changing {{title}} would require quite a lot of reasoning. 🙂

100% agree with you :) 
It’s very clear to the user that {{title}} is the note title :)   

> And, if we cannot change {{title}} to have a note_ prefix, I'm thinking that maybe we should not have a note_ prefix in any variable?

My opinion is that we use {{title}} the way that is and start a new “chapter” with the prefix `note_ ` for new variables.

Again, It’s very clear from the user that {{title}} is the note title :)    

> Would note_content then be explicitly content without YAML? What would be a name for a third variable that would give all content, including YAML?

Idea for the 3 variables:

- `yaml_content` —> content from yaml
- `note_content` —> content without yaml
- `all_content` Or `content ` —> the prefix `all` could remember the user of CMD/CRTL+A which is Select All  on Windows and MacOS

**Reply by Taitava** [reply]
> My opinion is that we use {{title}} the way that is and start a new “chapter” with the prefix `note_ ` for new variables.

I'll take it into consideration, but haven't decided yet. I'll have to see what other *note* related variable ideas will arise, to see the big picture.



> Idea for the 3 variables:
> 
>   - `yaml_content` —> content from yaml
>   - `note_content` —> content without yaml
>   - `all_content` Or `content ` —> the prefix `all` could remember the user of CMD/CRTL+A which is Select All  on Windows and MacOS

I think those are pretty good naming ideas.
 - `yaml_content`: I still need to think will the first word be *yaml*, *frontmatter*, *header*, or what. But the overall idea is good.
 - `note_content`: When seen in the group of these three variables, here I actually like the `note_` prefix, even if I'm otherwise uncertain about it. But here it would make sense.
 - `all_content` or `content`: I'd prefer `all_content`. And your aide-memoire* is quite good. 👍 
 
*) Huh, I've never seen that word before. I looked it up from a dictionary. I hope it makes sense here. 🙂

**Reply by FelipeRearden** [reply]
Great @Taitava !!!

Now is just a matter of choice :)

> *) Huh, I've never seen that word before. I looked it up from a dictionary. I hope it makes sense here. 🙂

As you can see since our first conversation, I barely know how to write the basic English ( many laughs). But I ask for help and I get it 
![Screen Shot 2021-10-02 at 11 14 01](https://user-images.githubusercontent.com/85180696/135720276-c29ec05d-f469-4048-85cf-fa9f36d38a7a.png)

And it makes total sense :) 

I wish you a fantastic Saturday. Have a great day!!!!!!!!

**Reply by Taitava** [reply]
Is that a plugin in your internet browser, or a native feature? Are you using Safari? That looks like a really easy to use tranlator/dictionary. I always open up a separate website and it takes time.

> As you can see since our first conversation, I barely know how to write the basic English ( many laughs).

I struggled many years with learning English (my mother language, Finnish, is so different than English). When I was a child, it was a pain in the school. But finally working many years in this field of IT (programming and other nerd stuff) taught me the vocabulary, and eventually I began to have the courage to speak face to face with foreign people in English without fearing what happens if I don't understand their accent, or what happens if I'm not able to bring to my mind a suitable word when I'm speaking. I had all these fears, but it's not the end of the world anymore if I encounter issues understanding other people's speach, or being understood myself. My English is not perfect still, and I probably use a wrong order of words in sentences, and wrong articles/prepositions. My point is: I can totally understand that everybody does not speak/write perfect English. It's a *lingua franca*, so there's quite many styles how different people write and speak English. 👍 

Have a great Saturday, too! 🌞

**Reply by FelipeRearden** [reply]
> Is that a plugin in your internet browser, or a native feature? Are you using Safari? That looks like a really easy to use tranlator/dictionary. I always open up a separate website and it takes time.

This is the built-in MacOS dictionary on Safari. A right mouse click open an option to open this pop-up window
- No Safari extension  :)
- There is an option to Dictation too.

![Screen Shot 2021-10-03 at 08 11 53](https://user-images.githubusercontent.com/85180696/135751112-234e1fb4-6252-4b6f-b9d3-7bbe3e803388.png)

> I struggled many years with learning English (my mother language, Finnish, is so different than English). When I was a child, it was a pain in the school. 

For Brazilians English is a challenge because its very different from Portuguese. But as you said Finnish and Portuguese are very different form English. 

All of this Obsidian way of life is good for me because I'm forced to write in English to talk to our community. In Economics we call this a positive externality :)

Have a great day @Taitava !!!!

**Reply by Taitava** [reply]
> In Economics we call this a positive externality :)

That's a great term! 🙂 Have a nice day, too! 🙂

**Reply by Taitava** [reply]
> ![Screen Shot 2021-10-02 at 11 14 01](https://user-images.githubusercontent.com/85180696/135720276-c29ec05d-f469-4048-85cf-fa9f36d38a7a.png)
> 

Hi @FelipeRearden! This is completely offtopic and not relevant to this plugin at all, but I remembered this screenshot of a translation feature in Safari. Do you know if there is a plugin for Obsidian that could provide similar translating features? Like so that I would select a word or sentence in Obsidian and maybe right click the selection and have some kind of "Look up" or "Translate" option in the context menu? I'm studying Swedish in school (among other studies not related to languages), and my workflow is to copy a few paragraphs in Swedish from an online book to Obsidian (by writing them manually, not copying and pasting, as I want to learn). Then I need to translate those paragraphs to Finnish. I try to do the translation by myself, but if I don't know a word, I'd like to have a super fast way to check it, rather than to open up a separate dictionary website.

If there is no plugin for this already, I could probably try to make one myself. Of course it would need to access external online services for doing the translation, so it's not so nice concerning privacy.

**Reply by FelipeRearden** [reply]
@Taitava just to let you know that I got your message.

Let just do a little search before answer you correctly. I think the answer is YES. But I wanna take a look first.

I get back to you :)

**Reply by FelipeRearden** [reply]
I found these two plugins:

https://github.com/TfTHacker/Obsidian-WordNet

https://github.com/phibr0/obsidian-dictionary


WordNet lets you import an Dictionary. I think is exactly what you want if you could import an Swedish to Finnish dictionary. 


obsidian-dictionary is not exactly what you want, but maybe if find a way to implement in you workflow

It is not exactly like Safari, but is close :)

---

As you know you can create a SC to invoke the plugin using Obsidian URI and paste the word (that is in clipboard) to the modal. 

When we have a {{selection:word}} you don't even have to copy the word to the clipboard :)

---

I think we have a third plugin related to translation. When I find it, I will write a message for you.



I'm pretty happy to have a chance to help you :)

Let keep talking

**Reply by Taitava** [reply]
Thanks @FelipeRearden! Sorry I missed your answer the last time I read this discussion.



> WordNet lets you import an Dictionary. I think is exactly what you want if you could import an Swedish to Finnish dictionary.

I took a quick glance at this and I'm not sure how it works. I think I should find a Swedish WordNet dictionary file from somewhere, and a quick Googling did not provide much results.

> obsidian-dictionary is not exactly what you want, but maybe if find a way to implement in you workflow

Maybe it might be. It just doesn't have Swedish, unfortunately, but I could try to look out what service it uses to get translations, and is there any possibility to add Swedish. I wonder why I did not find this plugin a few days ago when I searched with the word *dictionary*. Maybe I made a typo.

Thank you for these and have a nice day! 🙂

**Reply by FelipeRearden** [reply]
> Thanks @FelipeRearden! Sorry I missed your answer the last time I read this discussion.

No problem :) I was going to message you today just to remember :)
 
> > WordNet lets you import an Dictionary. I think is exactly what you want if you could import an Swedish to Finnish dictionary.

My hope was that you could import an Swedish to Finnish dictionary. The interface looks very good.
 
> I took a quick glance at this and I'm not sure how it works. I think I should find a Swedish WordNet dictionary file from somewhere, and a quick Googling did not provide much results.

I'm sorry if you didn't find yet.

> Maybe it might be. It just doesn't have Swedish, unfortunately, but I could try to look out what service it uses to get translations, and is there any possibility to add Swedish. I wonder why I did not find this plugin a few days ago when I searched with the word *dictionary*. Maybe I made a typo.

I find it by looking at the developer repository. Maybe it is not in the Community Store yet. 

I'm gonna keep looking for something. I will let you know if I find something interesting. 

I hope you find what you need for your workflow 🙏  

> Thank you for these and have a nice day! 🙂  

You too :)

---

Idea -> Could you create a new category in Discussion OFF-TOPIC for us to talk about this kind of thing ? 

This way we have a place to share with each other things that we find interesting. What you think?

**Reply by Taitava** [reply]
Nice! 🙂 The idea about an Offtopic category is good, so I created one! Thank you! 😎

**Reply by FelipeRearden** [reply]
Great @Taitava !!! Thank you :)

**Comment by Taitava** [comment]
And for the actual topic: I think it's a good time to implement this new variable in ~~`0.7.0`~~.

Edit 2021-12-10: Postponed due to [some problems](https://github.com/Taitava/obsidian-shellcommands/issues/77#issuecomment-952168545), but will be implemented later.

**Reply by FelipeRearden** [reply]
Thank you very much @Taitava ❤️ 🚀 👀  !!!!!!!

**Comment by catfist** [comment]
I'm now using this command to copy `{{content}}`.

```sh
cat "{{!file_path:relative}}" | sed -e '/^---$/,/^---$/d' -e '0,/^$/{//d}'
```

**Reply by Taitava** [reply]
Thanks, this is good! I feel this variable is one missing piece among the collection of file related variables. I need to refactor the varible system to make it asynchronous, which will allow creating this variable.

**Reply by Taitava** [reply]
I was able to make the variable system asynchronous. I'm working on these new varaibles now. 🙂

**Comment by Taitava** [comment]
# New variables for `0.16.0`
- `{{note_content}}`
- `{{file_content}}`
- `{{event_note_content}}`
- `{{event_file_content}}`

These are already implemented. I just need to test them more and write documentation.

# New variables later
- `{{yaml_content}}`
- `{{event_yaml_content}}`

I decided to implement these two variables later, because they'll need `Variable.isAvailable()` to be refectored to use `Promise`s, too. It's not an issue, as it's similar refactoring that I've already done in the variable system, but I just don't have time for it now, and I'd like to get `0.16.0` out asap. If a note file does not have a YAML section defined, `{{yaml_content}}` should make itself inavailable, making it possible for a user to fallback to a custom [default value](https://publish.obsidian.md/shellcommands/Variables/Default+values).

**Reply by Taitava** [reply]
# `0.16.0` is now released
Released variables:
-    {{note_content}}
-    {{file_content}}
-    {{event_note_content}}
-    {{event_file_content}}

Please comment here if you have any feedback regarding these new variables. 🙂

**Reply by Taitava** [reply]
# `{{yaml_content}}` and `{{event_yaml_content}}` were released in `0.18.0` a couple of days ago
I moved this discussion to _Ideas - Done & released_.
